### PR TITLE
Bounce the calendar selection to nearest valid available slot (TAM-119)

### PR DESCRIPTION
**Pull request by urshala**
Bounce the user's calendar selection to nearest valid slot if the
slot/selection of the user is not valid.

E.g. If the slot size is 1 hour and minimum reservation time is 2 hours
and user clicks on the last hour of calendar (say 17:00 i.e. The last
valid reservation is from 15:00 - 17:00).  Since the user clicked on 17:00,
the reservation 17:00 - 19:00 is not valid as it closes at 17:00, Thus, 
change the user's selection to start at 15:00 and notify the user of the
change.

bounce_to_nearest_hours

Refs TAM-119

### CODE REVIEW

File: app/i18n/messages/fi.json [hunk 1]
```diff
@@ -480,7 +480,7 @@
   "ResourceReservationCalendar.selectedDateLabel": "Valittu aika:",
   "ResourceReservationCalendar.selectedDateValue": "{date} klo {start} - klo {end} ({duration})",
   "ResourceReservationCalendar.selectedDateValueWithPrice": "{date} klo {start} - klo {end} ({duration}) yht {price}€",
-
+  "TimePickerCalendar.info.calendarSelectionMovedText": "Ajankohdan valinta on sovitettu \nkalenteriasetuksia vastaavaks",
```
Comment (tuomas777) [inline app/i18n/messages/fi.json]:
```suggestion
  "TimePickerCalendar.info.calendarSelectionMovedText": "Ajankohdan valinta on sovitettu kalenteriasetuksia vastaavaksi.",
```

File: src/common/calendar/TimePickerCalendar.js [hunk 2]
```diff
@@ -186,8 +190,19 @@ class TimePickerCalendar extends Component {
...
-      selectable = calendarUtils.getMinPeriodTimeRange(resource, selected.start, selected.end);
       // Make sure selected time will always bigger than min period
+      selectable = calendarUtils.getMinPeriodTimeRange(resource, selected.start, selected.end);
+
+      // Move the selection to nearest available slot
+      if (!this.isSelectionValid(selectable)) {
+        const nearestSlot = this.getNearestValidSlot(selectable, resource);
+        selectable = nearestSlot[0];
+        const nearestSlotIsValid = nearestSlot[1];
+        nearestSlotIsValid && createNotification(
```
Comment (tuomas777) [inline src/common/calendar/TimePickerCalendar.js]: Hmm, why not `if (nearestSlotIsValid) ...` ?

Comment (urshala) [reply src/common/calendar/TimePickerCalendar.js]: Wanted to save some lines with short circuit which would be same as `if (nearestSlotIsValid)...` .

File: app/i18n/messages/fi.json (hunk 1 above)
Comment (urshala) [reply app/i18n/messages/fi.json]: Fixed it :)

File: src/common/calendar/TimePickerCalendar.js (hunk 2 above)
Comment (tuomas777) [reply src/common/calendar/TimePickerCalendar.js]: Ok. A matter of opinion, but I don't really like replacing `if` and a function call with short-circuiting in this kind of places where it isn't proving any real benefits, and at least [some seem to agree with me](https://www.codereadability.com/dont-use-short-circuiting-when-you-want-an-if-statement/). But your PR so your call 🙂

**Review by tuomas777 — APPROVED** [review]
Looks good AFAICT 👍

File: src/common/calendar/TimePickerCalendar.js (hunk 2 above)
Comment (urshala) [reply src/common/calendar/TimePickerCalendar.js]: I agree :) 
Changed to `if (...)`

**Pull request merged**